plap/plap_provider.c

```diff
@@ -0,0 +1,446 @@
+
+static OpenVPNProvider *
+OpenVPNProvider_new(void)
+{
+    dmsg(L"Entry");
+
+    OpenVPNProvider *this = calloc(sizeof(*this), 1);
+
+    if (this)
+    {
+        this->lpVtbl = &icp_vtbl;
+        this->ref_count = 1; /* we free ourselves when this goes to zero */
+
+        dll_addref();
+    }
+
+    return this;
+}
+
+static void
+OpenVPNProvider_free(OpenVPNProvider *this)
+{
+    dmsg(L"Entry");
+
+    for (size_t i = 0; i < this->conn_count; ++i)
+    {
+        if (this->connections[i])
+        {
+            RELEASE((ICCPC*) this->connections[i]);
+        }
+    }
+    /* Destroy GUI threads and any associated data */
+    DeleteUI();
+
+    free(this);
+
+    dll_release();
+}
+
+/* Standard methods in every COM object inherited from IUnknown */
+static ULONG WINAPI
+AddRef(ICredentialProvider *this)
+{
+    OpenVPNProvider *op = (OpenVPNProvider *) this;
+
+    dmsg(L"ref_count after addref = %d", op->ref_count+1);
+
+    return InterlockedIncrement(&op->ref_count);
+}
+
+static ULONG WINAPI
+Release(ICredentialProvider *this)
+{
+    OpenVPNProvider *op = (OpenVPNProvider *) this;
+
+    ULONG count = InterlockedDecrement(&op->ref_count);
+
+    dmsg(L"ref_count after release = %d", op->ref_count);
+
+    if (op->ref_count == 0)
+    {
+        OpenVPNProvider_free(op); /* suicide -- equivalent of "delete this" */
+    }
+    return count;
+}
+
+/* In QueryInterface, return *ppv = pointer to the requested interface (riid),
+ * if we implement the interface.
+ * In our case we expect riid == IID_ICredentialProvider or IID_IUnknown
+ */
+static HRESULT WINAPI
+QueryInterface(ICredentialProvider *this, REFIID riid, void **ppv)
+{
+#ifdef DEBUG
+    debug_print_guid(riid, L"In Provider Queryinterface with iid = ");
+#endif
+
+    if (!ppv)
+    {
+        dmsg(L"ppv is NULL!");
+        return E_POINTER;
+    }
+    if (IsEqualIID(riid, &IID_IUnknown)
+        || IsEqualIID(riid, &IID_ICredentialProvider))
+    {
+        *ppv = this;
+        ADDREF(this);
+        return S_OK;
+    }
+    else
+    {
+        dmsg(L"unknown iid ignored");
+        *ppv = NULL;
+        return E_NOINTERFACE;
+    }
+}
+
+/*
+ * SetUsageScenario returns success for supported usages -- we support
+ * only PLAP.
+ *
+ * LogonUI calls this while initializing the provider, so we initialize
+ * our internal data structs, GUI related data, enumerate profiles,
+ * make a list of connections etc. This is done by calling
+ * CreateOVPNConnectionArray().
+ * After this, we should be ready to service calls for connection count
+ * and individual connection objects.
+ */
+static HRESULT WINAPI
+SetUsageScenario(ICredentialProvider *this,
+    CREDENTIAL_PROVIDER_USAGE_SCENARIO us, UNUSED DWORD flags)
+{
+    /* I think flags may be ignored for PLAP */
+
+    dmsg(L"cpus = %lu", us);
+
+    OpenVPNProvider *op = (OpenVPNProvider *) this;
+
+    op->usage_scenario = us; /* we need to keep this to pass on to connection objects */
```

Comment:
A PLAP provider can also be generic credential provider, so the dll registration in PLAP section alone doesn't say that we do not support any other usage scenarios. So we could get called with other usages.  However, we do not instantiate any connection instances unless usage is PLAP, so, right, we could avoid saving it here or checking it in ICredentialProviderCredential. But it can't hurt.